### ISSUE TITLE
Dash read built-in behavior change with 0.9 vs 0.7

### ISSUE THREAD

**Issue by shinji257** [lxc/lxcfs#33]
In 0.7 when running the command: read b c < /proc/uptime
variables $b and $c are populated with appropriate data.  In 0.9 they do not get data.  When running strace on lxcfs outside the container it shows that while in 0.7 they get data sent to them in 0.9 they are getting /0 sent instead.

When I test using bash (same command above) or using cat /proc/uptime as a placebo check they both act properly in 0.7 and 0.9 both giving data for uptime.

I'm putting this ticket in at Serge's request after posting on the mailing list.  Per his note it appears to be related to the implementation of direct_io which was done before the 0.8 release.

https://github.com/lxc/lxcfs/commit/7253e0a4b84fce836a2585ee5a982311163b6757

A strace on cat is trivial but one on dash is not (read is a builtin and strace won't do builtins) so I did that one for comparison sake.

Method used
stty -echo
cat | strace dash > /dev/null
Once that is done I had to type the command I wanted to strace blind when it paused waiting for input.  I enabled echo after it was all done.

https://pastebin.com/5T1xbHDH

**Comment by hallyn** [comment]
Hm, confirmed.

So this appears to be because dash is doing a partial read.  When I just do a

        fd = open(argv[1], O_RDONLY);
        ret = read(fd, &c, 1);

I get the same result dash gets. When I do

        ret = read(fd, buf, 50);

then it works.

**Comment by hallyn** [comment]
Two different things are going on.  First, the way snprintf is being used causes the last requested byte to be a \0.  Secondly, reads with offset are explicitly not being supported.
